src/qpdf/qpdf.cpp

```diff
@@ -400,11 +400,15 @@ void save_pdf(
         py::object ospath = py::module::import("os").attr("path");
         py::object samefile = ospath.attr("samefile");
         try {
-            if (samefile(filename, q.getFilename()).cast<bool>()) {
+            auto input_filename = q.getFilename();
+            // On Windows, if q was created over a stream, not a file path,
+            // q.getFilename() returns a string like "<_io.BytesIO object at 0x00ABCXYZ>".
+            auto input_is_stream = input_filename.front() == '<' && input_filename.back() == '>';
+            if (!input_is_stream && samefile(filename, input_filename).cast<bool>()) {
```

Comment:
I had to add this special case, otherwise `os.path.samefile` raises an `OSError` (not a `FileNotFoundError` catched below) due to `q.getFilename()` being an invalid filename